presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -71,78 +88,322 @@
+        if (node.getQuery() == newQuery) {
+            return node;
+        }
+
+        return new WithQuery(node.getName(), processSameType(node.getQuery()), node.getColumnNames());
+    }
+
+    private <T extends Node> List<T> processNodes(List<T> nodes)
+    {
+        List<T> newNodes = new ArrayList<>();
+        boolean listsIdentical = true;
+
+        for (T node : nodes) {
+            T newNode = processSameType(node);
+            if (node != newNode) {
+                listsIdentical = false;
+            }
+            newNodes.add(newNode);
+        }
+        return listsIdentical ? nodes : newNodes;
+    }
+
+    private <T extends Node> T processSameType(T node)
+    {
+        return (T) process(node);
+    }
+
+    private QuerySpecification rewriteQuerySpecificationIfCompatible(QuerySpecification querySpecification, Table baseTable)
+    {
+        Optional<String> errorMessage =
+                MaterializedViewRewriteQueryShapeValidator.validate(querySpecification);
+
+        if (errorMessage.isPresent()) {
+            log.info(errorMessage.get());
+            return querySpecification;
+        }
+
+        List<QualifiedObjectName> referencedMaterializedViews = metadata.getReferencedMaterializedViews(
+                session,
+                createQualifiedObjectName(session, baseTable, baseTable.getName()));
+
+        // TODO: Select the most compatible and efficient materialized view for query rewrite optimization https://github.com/prestodb/presto/issues/16431
+        // TODO: Refactor query optimization code https://github.com/prestodb/presto/issues/16759
+
+        for (QualifiedObjectName materializedViewName : referencedMaterializedViews) {
+            QuerySpecification rewrittenQuerySpecification = getRewrittenQuerySpecification(metadata, materializedViewName, querySpecification);
+
+            if (rewrittenQuerySpecification != querySpecification) {
+                if (isMaterializedViewDataConsistencyEnabled(session)) {
+                    // TODO: We should be able to leverage this information in the StatementAnalyzer as well.
+                    MaterializedViewStatus materializedViewStatus = metadata.getMaterializedViewStatus(session, materializedViewName);
+                    if (materializedViewStatus.isPartiallyMaterialized() || materializedViewStatus.isFullyMaterialized()) {
+                        session.getRuntimeStats().addMetricValue(OPTIMIZED_WITH_MATERIALIZED_VIEW_SUBQUERY_COUNT, NONE, 1);
+                        return rewrittenQuerySpecification;
+                    }
+                    session.getRuntimeStats().addMetricValue(MANY_PARTITIONS_MISSING_IN_MATERIALIZED_VIEW_COUNT, NONE, 1);
+                }
+                else {
+                    session.getRuntimeStats().addMetricValue(OPTIMIZED_WITH_MATERIALIZED_VIEW_SUBQUERY_COUNT, NONE, 1);
+                    return rewrittenQuerySpecification;
+                }
+            }
```

Comment:
nit - too many nested `if`s:
```suggestion
if (rewrittenQuerySpecification == querySpecification) {
	continue;
}

if (!isMaterializedViewDataConsistencyEnabled(session)) {
	session.getRuntimeStats().addMetricValue(OPTIMIZED_WITH_MATERIALIZED_VIEW_SUBQUERY_COUNT, NONE, 1);
	return rewrittenQuerySpecification;
}

// TODO: We should be able to leverage this information in the StatementAnalyzer as well.
MaterializedViewStatus materializedViewStatus = metadata.getMaterializedViewStatus(session, materializedViewName);
if (materializedViewStatus.isPartiallyMaterialized() || materializedViewStatus.isFullyMaterialized()) {
    session.getRuntimeStats().addMetricValue(OPTIMIZED_WITH_MATERIALIZED_VIEW_SUBQUERY_COUNT, NONE, 1);
    return rewrittenQuerySpecification;
}
session.getRuntimeStats().addMetricValue(MANY_PARTITIONS_MISSING_IN_MATERIALIZED_VIEW_COUNT, NONE, 1);
```